pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -0,0 +1,188 @@
+        self.save_weights_only = save_weights_only
+        self.period = period
+        self.epochs_since_last_check = 0
+        self.prefix = prefix
+        self.best_k_models = {}
+        # {filename: monitor}
+        self.kth_best_model = ''
+        self.best = 0
+        self.save_function = None
+
+        if mode not in ['auto', 'min', 'max']:
+            warnings.warn(
+                f'ModelCheckpoint mode {mode} is unknown, '
+                'fallback to auto mode.', RuntimeWarning)
+            mode = 'auto'
+
+        if mode == 'min':
+            self.monitor_op = np.less
+            self.kth_value = np.Inf
+            self.mode = 'min'
+        elif mode == 'max':
+            self.monitor_op = np.greater
+            self.kth_value = -np.Inf
+            self.mode = 'max'
+        else:
+            if 'acc' in self.monitor or self.monitor.startswith('fmeasure'):
+                self.monitor_op = np.greater
+                self.kth_value = -np.Inf
+                self.mode = 'max'
+            else:
+                self.monitor_op = np.less
+                self.kth_value = np.Inf
+                self.mode = 'min'
+
+    def _del_model(self, filepath):
+        try:
+            shutil.rmtree(filepath)
+        except OSError:
+            os.remove(filepath)
+
+    def _save_model(self, filepath):
+        # make paths
+        os.makedirs(os.path.dirname(filepath), exist_ok=True)
+
+        # delegate the saving to the model
+        assert self.save_function is not None, ".save_function() not set"
+        self.save_function(filepath)
+
+    def check_monitor_top_k(self, current):
+        less_than_k_models = len(self.best_k_models) < self.save_top_k
+        if less_than_k_models:
+            return True
+        return self.monitor_op(current, self.best_k_models[self.kth_best_model])
+
+    def on_validation_end(self):
+        logs = self.trainer.callback_metrics
+        epoch = self.trainer.current_epoch
+        self.epochs_since_last_check += 1
+
+        if self.save_top_k == 0:
+            # no models are saved
+            return
+        if self.epochs_since_last_check >= self.period:
+            self.epochs_since_last_check = 0
+            filepath = f'{self.filepath}/{self.prefix}_ckpt_epoch_{epoch}.ckpt'
+            version_cnt = 0
+            while os.path.isfile(filepath):
+                # this epoch called before
+                filepath = f'{self.filepath}/{self.prefix}_ckpt_epoch_{epoch}_v{version_cnt}.ckpt'
+                version_cnt += 1
+
+            if self.save_top_k != -1:
+                current = logs.get(self.monitor)
+
+                if current is None:
+                    warnings.warn(
+                        f'Can save best model only with {self.monitor} available,'
+                        ' skipping.', RuntimeWarning)
+                else:
+                    if self.check_monitor_top_k(current):
+
+                        # remove kth
```

Comment:
it would be better readable to cut-out this section to separate (protected) method...